alternative-1-iterate-generic.Rmd

```diff
@@ -78,81 +82,109 @@ iterate.SquaresSequence <- function(x, state = NULL) {
 
 ```{r}
-iterate.python.builtin.object <- function(x, state = NULL) {
-  if(is.null(state))
-    state <- reticulate::as_iterator(x)
+SampleSequence <- function(max) {
+  structure(
+    list(max = max),
+    class = "SampleSequence"
+  )
+}
 
-  sentinal <- environment()
-  value <- reticulate::iter_next(state, completed = sentinal)
-  if(identical(value, sentinal))
-    NULL
-  else
-    list(value, state)
+iterate.SampleSequence <- function(x, state = NULL) {
+  if(is.null(state))
+    state <- 0L
+  if (state > x$max)
+    return(NULL)
+  
+  val <- abs(rnorm(1))
+  list(
+    value = val, 
+    state = state + val
+  )
 }
 
-for(x in reticulate::r_to_py(1:3))
+for (x in SampleSequence(2)) {
   print(x)
+}
 ```
 
-The above is a narrow but flexible change (Only one new symbol would be
-introduced, `base::iterate()`). Since `iterate()` is a new generic, it
-would allows for changes in behavior to be made with intention
-individually for each class type.
+## Extensions
 
-## Optional extensions
+### Support for S3 classes
 
-### `iterate.function()`
+A simple modification to the default iterator (from `.subset2()` to
+`[[`) makes it possible to selectively change the behavior of key S3
+classes:
 
-An optional extension might be for the `base` namespace to also provide
-a `iterate()` method for functions. This would allow users to pass
-functions directly to `for`.
+```{r}
+iterate.POSIXt <- function(x, state = NULL) {
+  if(is.null(state)) # start of iteration
+    state <- 1L
+  if(state > length(x)) # end of iteration
+    return(NULL)
+  list(
+    value = x[[state]], 
+    state = state + 1L
+  )
+}
+iterate.factor <- iterate.POSIXt
 
-In this scenario, user defined functions would be responsible for
-indicating when an iterator function is exhausted, either by signaling a
-condition or returning a sentinal (or running forever). For example,
-with a sentinal based approach:
+dt <- .POSIXct(c(1, 2))
+for (x in dt) {
+  str(x)
+}
 
-```{r}
-iterate.function <- function(x, state = NULL) {
-  if (is.null(state)) # start of iteration
-    state <- x
-  
-  value <- state()
-  if (identical(value, IteratorExhaustedSentinal))
-    NULL
-  else
-    list(value, state)
+for (x in as.POSIXlt(dt)) {
+  str(x)
 }
 
-IteratorExhaustedSentinal <- new.env(parent = emptyenv())
+for (x in factor(c("x", "y"))) {
+  str(x)
+}
 ```
 
-Then users could write code like this:
+### `iterate.environment()`
+
+An environment method would allow users to pass environments to `for`.
+Environments are mutable so there are some challenges if the environment
+changes during implementation. This implementation works by keeping
+track of which elements have already been processed.
 
 ```{r}
-SquaresSequenceIterator <- function(from = 1, to = 10) {
-  force(from); force(to)
-  function() {
-    if(from > to)
-      return(IteratorExhaustedSentinal)
-    on.exit(from <<- from + 1L)
-    from ^ 2
+iterate.environment <- function(x, state = NULL) {
```

Comment:
I changed the implementation here — I think this becomes a nice example of where state might be something other than a number.